miner/scroll_worker.go

```diff
@@ -751,57 +876,69 @@ func copyReceipts(receipts []*types.Receipt) []*types.Receipt {
 	return result
 }
 
-func (w *worker) onTxFailingInPipeline(txIndex int, tx *types.Transaction, err error) bool {
+func (w *worker) onTxFailing(txIndex int, tx *types.Transaction, err error) bool {
 	if !w.isRunning() {
 		return false
 	}
 
-	writeTrace := func() {
-		var trace *types.BlockTrace
-		var errWithTrace *pipeline.ErrorWithTrace
-		if w.config.StoreSkippedTxTraces && errors.As(err, &errWithTrace) {
-			trace = errWithTrace.Trace
+	if errors.Is(err, ccc.ErrBlockRowConsumptionOverflow) {
+		if txIndex > 0 {
+			if !tx.IsL1MessageTx() {
+				// prioritize overflowing L2 message as the first txn next block
+				// no need to prioritize L1 messages, they are fetched in order
+				// and processed first in every block anyways
+				w.prioritizedTx = &prioritizedTransaction{
+					blockNumber: w.current.header.Number.Uint64() + 1,
+					tx:          tx,
+				}
+			}
+			return true
 		}
-		rawdb.WriteSkippedTransaction(w.eth.ChainDb(), tx, trace, err.Error(),
-			w.currentPipeline.Header.Number.Uint64(), nil)
-	}
 
-	switch {
-	case errors.Is(err, core.ErrGasLimitReached) && tx.IsL1MessageTx():
-		// If this block already contains some L1 messages try again in the next block.
-		if txIndex > 0 {
-			break
+		// first txn overflowed the circuit, skip
+		w.skipTransaction(tx, err)
+	} else if tx.IsL1MessageTx() {
+		if errors.Is(err, ErrUnexpectedL1MessageIndex) {
+			log.Warn(
+				"Unexpected L1 message queue index in worker", "got", tx.AsL1MessageTx().QueueIndex,
+			)
+			return false
+		} else if txIndex > 0 {
+			// If this block already contains some L1 messages try again in the next block.
+			return false
 		}
-		// A single L1 message leads to out-of-gas. Skip it.
-		queueIndex := tx.AsL1MessageTx().QueueIndex
-		log.Info("Skipping L1 message", "queueIndex", queueIndex, "tx", tx.Hash().String(), "block",
-			w.currentPipeline.Header.Number, "reason", "gas limit exceeded")
-		writeTrace()
-		l1TxGasLimitExceededCounter.Inc(1)
 
-	case errors.Is(err, core.ErrInsufficientFunds):
+		queueIndex := tx.AsL1MessageTx().QueueIndex
+		log.Warn("Skipping L1 message", "queueIndex", queueIndex, "tx", tx.Hash().String(), "block",
+			w.current.header.Number, "reason", err)
+		rawdb.WriteSkippedTransaction(w.eth.ChainDb(), tx, nil, err.Error(),
+			w.current.header.Number.Uint64(), nil)
+		w.current.nextL1MsgIndex = queueIndex + 1
+		l1SkippedCounter.Inc(1)
+	} else if errors.Is(err, core.ErrInsufficientFunds) {
 		log.Trace("Skipping tx with insufficient funds", "tx", tx.Hash().String())
 		w.eth.TxPool().RemoveTx(tx.Hash(), true)
+	}
```

Comment:
What about adding an else branch to log the unexpected cases? e.g. L2 transactions not due to the reasons in the above conditional branches.

Reply:
I am worried about spamming the logs. But I can still add it if you think it will be useful

Reply:
I may misunderstand this. Won't it be impossible to go into the `else` branch (i.e. only the expected case) as designed?

Reply:
I see. then no problem.